shell/common/api/electron_api_clipboard_gtk.cc

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2020 Microsoft, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/common/api/electron_api_clipboard.h"
+
+#include <gtk/gtk.h>
+
+namespace electron {
+
+namespace api {
+
+namespace {
+
+void ClipboardGet(GtkClipboard* clipboard,
+                  GtkSelectionData* selection,
+                  guint info,
+                  gchar** uris) {
+  DCHECK_EQ(info, 0u);
+  gtk_selection_data_set_uris(selection, uris);
+}
+
+void ClipboardClear(GtkClipboard* clipboard, gchar** uris) {
+  g_strfreev(uris);
+}
+
+}  // namespace
+
+std::vector<base::FilePath> Clipboard::ReadFilePaths() {
+  std::vector<base::FilePath> result;
+  GtkClipboard* clipboard = gtk_clipboard_get(GDK_SELECTION_CLIPBOARD);
+  gchar** uris = gtk_clipboard_wait_for_uris(clipboard);
```

Comment:
This seems like a dangerous function to call:

> This function waits for the data to be received using the main loop, so events, timeouts, etc, may be dispatched during the wait.

This means V8 could be re-entered (e.g. through an emitted event) while this method is running, which is undesirable IMO.

Reply:
This is unfortunately also Chromium does, doing a nested run loop until data is received:
https://source.chromium.org/chromium/chromium/src/+/master:ui/base/clipboard/clipboard_x11.cc;l=354;drc=28d989a1fc83edb779c034cad48572775696bc66;bpv=0;bpt=1
https://source.chromium.org/chromium/chromium/src/+/master:ui/base/x/selection_requestor.cc;l=236;drc=510d3cc7ecba433b10d6854f3623ed4ca8fb72d7

For X11 reading clipboard is an async operation so this behavior can not be avoided.

Reply:
Chromium doesn't run V8 in the main process though, so reentry is much safer there. Is there some other way we could hook into the event loop here? Could we post a task?

Reply:
Electron's clipboard APIs use Chromium's clipboard implementation, though Chromium does not run V8 in the main process, Electron does, so every time Electron app uses clipboard API on Linux, it has the reentry problem when waiting for X11 clipboard to respond.

I think using GTK's sync clipboard API has no difference from using Chromium's clipboard implementations.